### PR TITLE
[FIX] account,sale,pos,*: branch user access

### PR DESCRIPTION
**Pull request by yosuanicolaus**
When a user only has access to a branch of a company, there is a lot of places where AccessError pops up because of how the domain was previously set on them.

This commit aims to be a part of frequent small PR in the future to fix these problems.

These PRs fixes:

- Unaccessible parent company's fields (now accessible) in 
	- Sales/Purchase Orders/Quotation form: product variant, product (product_template), and tax
	- Account Move (invoice/bill/etc) line, when filling product
- Access Error as user with only access to branch when:
	- opening Accounting reports (BS / PL)
	- opening PoS dashboard
	- trying to add product when creating new invoice
- Improves Journal access rules. Instead of making all parent's journal public, base the user access on: (explicitly granted access from user allowed companies) or (all self&parent's journal AND journal type is not Bank/Cash) 
- Improves styling on report's Journal dropdown button's style (in enterprise PR)

community-PR: https://github.com/odoo/odoo/pull/140078
enterprise-PR: https://github.com/odoo/enterprise/pull/49820

[task-id: 3551305](https://www.odoo.com/web?debug=#id=3551305&cids=1&menu_id=4720&action=333&active_id=967&model=project.task&view_type=form)

### CODE REVIEW

File: odoo/addons/base/models/ir_rule.py [hunk 1]
```diff
@@ -200,7 +200,7 @@ def _make_access_error(self, operation, records):
 
         if not self.user_has_groups('base.group_no_one') or not self.env.user.has_group('base.group_user'):
             records.invalidate_recordset()
-            return AccessError(f"{operation_error}\n\n{resolution_info}")
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_rule.py:203]: Is it intentional ? It isn't legit at all :thinking:

Comment (yosuanicolaus) [reply odoo/addons/base/models/ir_rule.py:203]: Sorry, no, that was just for debugging, and the PR is not fully done.
I didn't expect robodoo to ask review from so many people :laughing:

**Review by epictete — COMMENTED** [review]
Hello @yosuanicolaus 
Could you please split your code into dedicated commits, each addressing a specific use case/bug, and provide steps to reproduce each?

**Comment by yosuanicolaus** [comment]
@epictete I've split both PRs to multiple commit, each with long message for more context about why I'm doing it this way. :smile:

I'm not sure if I can change the `sudo` part though. I know it looks ugly, but they are because of how the rule was set up for branch company.

File: addons/sale/models/sale_order_line.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ class SaleOrderLine(models.Model):
     product_id = fields.Many2one(
         comodel_name='product.product',
         string="Product",
-        change_default=True, ondelete='restrict', check_company=True, index='btree_not_null',
+        change_default=True, ondelete='restrict', index='btree_not_null',
```
Comment (Levizar) [inline addons/sale/models/sale_order_line.py:84]: I think we should add the parent check in the domain just below.
Without that, I think removing this check risk to display products from a completely unrelated company.

Did you check whether that was the case or not?

File: addons/sale/tests/test_sale_order.py [hunk 3]
```diff
@@ -629,6 +629,55 @@ def test_cannot_assign_tax_of_mismatch_company(self):
...
             sol.tax_id = tax_b
 
+    def test_assign_tax_multi_company(self):
+        root_company = self.env['res.company'].create({'name': 'B0 company'})
+        root_company.write({'child_ids': [
+            Command.create({'name': 'B1 company'}),
+            Command.create({'name': 'B2 company'}),
+        ]})
+
+        country = self.env['res.country'].search([])[0]
```
Comment (Levizar) [inline addons/sale/tests/test_sale_order.py:639]:
```suggestion
        country = self.env['res.country'].search([], limit=1)
```
Would be nice to change it on line 598 too (The test above)

**Comment by Levizar** [comment]
Forgot to write it but Good Job OFC! :100:

**Comment by Levizar** [comment]
Also forgot to write it: all the task-id in the commit message are wrong because they point to an already deployed task.
Could you update them?

File: addons/account/models/account_payment.py [hunk 4]
```diff
@@ -459,17 +459,19 @@ def _compute_available_journal_ids(self):
...
             if pay.payment_type == 'inbound':
                 pay.available_journal_ids = journals.filtered(
-                    lambda j: j.company_id == pay.company_id and j.inbound_payment_method_line_ids.ids != []
+                    lambda j: j.inbound_payment_method_line_ids.ids != []
                 )
             else:
                 pay.available_journal_ids = journals.filtered(
-                    lambda j: j.company_id == pay.company_id and j.outbound_payment_method_line_ids.ids != []
+                    lambda j: j.outbound_payment_method_line_ids.ids != []
                 )
```
[the remark this answers is not in the document]
Comment (yosuanicolaus) [reply addons/account/models/account_payment.py:475]: thanks for the suggestions! I'll use the 1st one as it's easier to understand ^^

File: addons/sale/models/sale_order_line.py (hunk 2 above)
Comment (yosuanicolaus) [reply addons/sale/models/sale_order_line.py:84]: We have a constrain check for product_id company already, in
- [here](https://github.com/odoo/odoo/pull/140078/files#diff-86168e4c37e9a7c1437998990c2aa4e1690d16605812f57a81172936377b9c50R157-R168) for purchase
- [here](https://github.com/odoo/odoo/pull/140078/files#diff-695321b9df7129604a59fbc9bff34f97ba90b3394aa1166ca24e36b27389ae82R668-R679) for sale

Comment (Levizar) [reply addons/sale/models/sale_order_line.py:84]: This is preventing to save the invoice with those product but not to display them.
But I tested in the UI and the products don't show up so it's ok! :ok_hand:

**Review by Levizar — APPROVED** [review]
@robodoo r+ rebase-ff

as already said: well done!

**Pull request closed without merging**